redis/tests/test_cache.rs

```diff
@@ -0,0 +1,438 @@
+fn test_cache_basic(#[case] runtime: RuntimeType, #[values(true, false)] test_with_optin: bool) {
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+    block_on_all(
+        async move {
+            let cache_config = if test_with_optin {
+                CacheConfig::new().set_mode(redis::caching::CacheMode::OptIn)
+            } else {
+                CacheConfig::default()
+            };
+            let mut con = ctx.async_connection_with_cache_config(cache_config).await?;
+            let val: Option<String> = get_cmd("GET", test_with_optin)
+                .arg("key_1")
+                .query_async(&mut con)
+                .await
+                .unwrap();
+            assert_eq!(val, None);
+            assert_hit(&con, 0);
+            assert_miss(&con, 1);
+
+            let val: Option<String> = get_cmd("GET", test_with_optin)
+                .arg("key_1")
+                .query_async(&mut con)
+                .await
+                .unwrap();
+            assert_eq!(val, None);
+            // key_1's value should be returned from cache even if it doesn't exist in server yet.
+            assert_hit(&con, 1);
+            assert_miss(&con, 1);
+
+            let _: () = get_cmd("SET", test_with_optin)
+                .arg("key_1")
+                .arg("1")
+                .query_async(&mut con)
+                .await
+                .unwrap();
+            sleep(Duration::from_millis(50).into()).await; // Give time for push message to be received after invalidating key_1.
+            assert_hit(&con, 1);
+            assert_miss(&con, 1);
+            assert_invalidate(&con, 1);
+
+            let val: String = get_cmd("GET", test_with_optin)
+                .arg("key_1")
+                .query_async(&mut con)
+                .await
+                .unwrap();
+            assert_eq!(val, "1");
+            // After invalidating key_1, now it misses the key from cache
+            assert_hit(&con, 1);
+            assert_miss(&con, 2);
+            assert_invalidate(&con, 1);
+            Ok::<_, RedisError>(())
+        },
+        runtime,
+    )
+    .unwrap();
+}
+
+#[rstest]
+#[case::tokio(RuntimeType::Tokio)]
+#[cfg_attr(feature = "async-std-comp", case::async_std(RuntimeType::AsyncStd))]
+fn test_cache_mget(#[case] runtime: RuntimeType) {
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(
+        async move {
+            let mut con = ctx.async_connection_with_cache().await?;
+            // Caching must work with both MGET and GET
+            let _: Vec<Option<String>> = redis::cmd("MGET")
+                .arg("key_1")
+                .arg("key_2")
+                .query_async(&mut con)
+                .await?;
+            assert_hit(&con, 0);
+            assert_miss(&con, 2);
+
+            let _: Vec<Option<String>> = redis::cmd("MGET")
```

Comment:
can this check for keys 1,2, &3, but in the order of 1,3,2?
Just to verify that the caching mechanism is not in some way order dependent

Reply:
Changed the order also added SET and asserts for values.